gcp/src/main/java/org/apache/iceberg/gcp/biglake/BigLakeCatalog.java

```diff
@@ -0,0 +1,398 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigLakeCatalog.class);
+
+  // The name of this Iceberg catalog plugin.
+  private String name;
+  private Map<String, String> properties;
+  private FileIO io;
+  private Object conf;
+
+  private String projectId;
+  private String region;
+  // BLMS catalog ID and fully qualified name.
+  private String catalogId;
+  private CatalogName catalogName;
+  private BigLakeClient client;
+
+  private CloseableGroup closeableGroup;
+
+  // Must have a no-arg constructor to be dynamically loaded
+  // initialize(String name, Map<String, String> properties) will be called to complete
+  // initialization
+  public BigLakeCatalog() {}
+
+  @Override
+  public void initialize(String initName, Map<String, String> initProperties) {
+    BigLakeClient initClient;
+    try {
+      // CatalogProperties.URI specifies the endpoint of BigLake API.
+      // TODO: to add more auth options of the client. Currently it uses default auth
+      // (https://github.com/googleapis/google-cloud-java#application-default-credentials)
+      // that works on GCP services (e.g., GCE, GKE, Dataproc).
+      initClient =
+          new BigLakeClient(
+              initProperties.getOrDefault(CatalogProperties.URI, DEFAULT_BIGLAKE_SERVICE_ENDPOINT));
+    } catch (IOException e) {
+      throw new ServiceFailureException(e, "Creating BigLake client failed");
+    }
+
+    initialize(initName, initProperties, initClient);
+  }
+
+  @VisibleForTesting
+  void initialize(String initName, Map<String, String> initProperties, BigLakeClient initClient) {
+    this.name = initName;
+    this.properties = ImmutableMap.copyOf(initProperties);
+
+    Preconditions.checkArgument(initClient != null, "BigLake client must not be null");
+    this.client = initClient;
+
+    Preconditions.checkArgument(
+        properties.containsKey(GCPProperties.PROJECT_ID), "GCP project ID must be specified");
+    this.projectId = properties.get(GCPProperties.PROJECT_ID);
+
+    Preconditions.checkArgument(
+        properties.containsKey(GCPProperties.REGION), "GCP region must be specified");
+    this.region = properties.get(GCPProperties.REGION);
+
+    // Users can specify the BigLake catalog ID, otherwise catalog plugin name will be used.
+    // For example, "spark.sql.catalog.<plugin-name>=org.apache.iceberg.spark.SparkCatalog"
+    // specifies the plugin name "<plugin-name>".
+    this.catalogId = properties.getOrDefault(GCPProperties.BIGLAKE_CATALOG_ID, initName);
+    this.catalogName = CatalogName.of(projectId, region, catalogId);
+
+    String ioImpl =
+        properties.getOrDefault(CatalogProperties.FILE_IO_IMPL, ResolvingFileIO.class.getName());
+    this.io = CatalogUtil.loadFileIO(ioImpl, properties, conf);
+
+    this.closeableGroup = new CloseableGroup();
+    closeableGroup.addCloseable(io);
+    closeableGroup.setSuppressCloseFailure(true);
```

Comment:
should the client also be added to the closeable group? Right now `BigLakeClient` doesn't have a `close()` method but I think it should implement `AutoCloseable`, because the `MetastoreServiceClient` in  `BigLakeClient` is also `AutoCloseable`

Reply:
Done.